File: 백준/Gold/12904. A와 B/A와 B.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String s = br.readLine();
        String t = br.readLine();

        String sReverse = new StringBuilder(s).reverse().append("B").toString();

        int sLength = s.length();
        int tLength = t.length();

        int[] countArr = new int[tLength];
        if (t.charAt(0) == 'B') countArr[0] = 1;
        for (int i = 1; i < tLength; i++) {
            if (t.charAt(i) == 'B') countArr[i]++;
            countArr[i] += countArr[i - 1];
        }

        int answer = 0;
        for (int i = 0; i <= tLength - sLength; i++) {
            if (t.substring(i, i + sLength).equals(s)) {
                if (!check_B(i - 1, i + sLength, tLength, countArr)) continue;
                if (check_A(i - 1, t)) {
                    answer = 1;
                    break;
                }
            }
        }

        if (answer == 0) {
            for (int i = 0; i < tLength - sLength; i++) {
                if (t.substring(i, i + sLength + 1).equals(sReverse)) {
                    if (!check_B(i - 1, i + sLength + 1, tLength, countArr)) continue;
                    answer = 1;
                    break;
                }
            }
        }

        System.out.println(answer);
    }

    public static boolean check_A(int leftEnd, String t) {
        if (leftEnd >= 0) {
            for (int i = leftEnd; i >= 0; i--) {
                if (t.charAt(i) == 'A') {
                    if (i != leftEnd && t.charAt(i + 1) == 'B') break;
                    return false;
                }
            }
        }
        return true;
    }

    public static boolean check_B(int leftEnd, int rightStart, int tLength, int[] countArr) {
        if (leftEnd < 0) {
            return  countArr[tLength - 1] - countArr[rightStart - 1] == 0;
        }
        return countArr[leftEnd] == countArr[tLength - 1] - countArr[rightStart - 1];
    }

}

File: 백준/Gold/12904. A와 B/A와 B.py
def check_A(left_end, T):
    if left_end >= 0:
        for i in range(left_end, -1, -1):
            if T[i] == "A":
                if i != left_end and T[i + 1] == "B":
                    break
                return False
    return True

def check_B(i, s, t, count_B):
    if i < 0:
        return count_B[t - 1] - count_B[i + s] == 0
    return count_B[i] == count_B[t - 1] - count_B[i + s]

def solution():
    S = input()
    T = input()

    S_reverse = S[::-1] + "B"

    sl = len(S)
    tl = len(T)

    count_B = [0] * tl
    if T[0] == "B": count_B[0] = 1
    for i in range(1, tl):
        if T[i] == "B":
            count_B[i] += 1
        count_B[i] += count_B[i - 1]

    answer = 0
    for i in range(tl - sl + 1):
        if T[i:i + sl] == S:
            if not check_B(i - 1, sl, tl, count_B):
                continue

            if check_A(i - 1, T):
                return 1

    for i in range(tl - sl):
        if T[i:i + sl + 1] == S_reverse:
            if not check_B(i - 1, sl + 1, tl, count_B):
                continue
            return 1
    return 0


print(solution())